README.md
# mrg-random
This package contains pure haskell implementation of some
MRG (Multiple Recursive Generator) pseudo random number generators.

MRG PRNGs(`MRG32k3a`, `MRG63k3a`) have good features as following.

  * Long period (`MRG32k3a`: ≈ 2¹⁹¹, `MRG63k3a`: ≈ 2³⁷⁷).
  * Generate high quality random numbers following U(0,1).
  * Small state vector (only 6 elements of `Double` / `Int64`).
  * Able to extract substream at small cost.



CHANGELOG.md
# Revision history for mrg-random

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/System/Random/MRG.hs
{- -*- mode: haskell; coding: utf-8-unix -*-  -}
{-# LANGUAGE BangPatterns, CPP, DeriveDataTypeable, FlexibleContexts,
    ScopedTypeVariables, TypeFamilies #-}

module System.Random.MRG
    (
      MRGen
    , initialize

    , uniform

    , Seed
    , fromSeed
    , save
    , restore

    , jump
    ) where

-- import System.Random
import Data.Typeable (Typeable)
import Data.Word (Word32,Word64)
import Data.List (unfoldr)

import System.Random.MRG.Internal

data MRGen = MRGen Double Double Double Double Double Double

norm :: Double
norm = 2.328306549295728e-10
{-# INLINE norm #-}

m1 :: Word64
m1 = 4294967087
{-# INLINE m1 #-}

m1f :: Double
m1f = fromIntegral m1
{-# INLINE m1f #-}

m2 :: Word64
m2 = 4294944443
{-# INLINE m2 #-}

m2f :: Double
m2f = fromIntegral m2
{-# INLINE m2f #-}

a12 :: Word64
a12 = 1403580
{-# INLINE a12 #-}

a12f :: Double
a12f = fromIntegral a12
{-# INLINE a12f #-}

a13n :: Word64
a13n = 810728
{-# INLINE a13n #-}

a13nf :: Double
a13nf = fromIntegral a13n
{-# INLINE a13nf #-}

a21 :: Word64
a21 = 527612
{-# INLINE a21 #-}

a21f :: Double
a21f = fromIntegral a21
{-# INLINE a21f #-}

a23n :: Word64
a23n = 1370589
{-# INLINE a23n #-}

a23nf :: Double
a23nf = fromIntegral a23n
{-# INLINE a23nf #-}

floorInt :: Double -> Int
floorInt = floor

mrg32k3a_genRand :: MRGen -> (Double,MRGen)
mrg32k3a_genRand (MRGen s10 s11 s12 s20 s21 s22)
  = (v, MRGen s11 s12 t1 s21 s22 t2)
  where p1 = a12f * s11 - a13nf * s10
        q1 = floorInt (p1 / m1f)
        r1 = p1 - fromIntegral q1 * m1f
        !t1 = if r1 < 0.0 then r1 + m1f else r1
        p2 = a21f * s22 - a23nf * s20
        q2 = floorInt (p2 / m2f)
        r2 = p2 - fromIntegral q2 * m2f
        !t2 = if r2 < 0.0 then r2 + m2f else r2
        !v = if t1 <= t2 then (t1 - t2 + m1f) * norm else (t1 - t2) * norm
{-# INLINE mrg32k3a_genRand #-}

initialize :: (Integral a) => a -> MRGen
initialize seed = MRGen s1 s1 s1 s2 s2 s2
  where s1 = fromIntegral $ fromIntegral seed `mod` m1
        s2 = fromIntegral $ fromIntegral seed `mod` m2
{-# INLINE initialize #-}

uniform :: MRGen -> (Double,MRGen)
uniform gen = mrg32k3a_genRand gen
{-# INLINE uniform #-}

newtype Seed = Seed { fromSeed :: (Word32,Word32,Word32,Word32,Word32,Word32) } deriving (Eq, Show, Typeable)

save :: MRGen -> Seed
save (MRGen s10 s11 s12 s20 s21 s22) = Seed (t1,t2,t3,t4,t5,t6)
  where t1 = floor s10
        t2 = floor s11
        t3 = floor s12
        t4 = floor s20
        t5 = floor s21
        t6 = floor s22
{-# INLINE save #-}

restore :: Seed -> MRGen
restore (Seed (t1,t2,t3,t4,t5,t6)) = MRGen s10 s11 s12 s20 s21 s22
  where s10 = fromIntegral $ t1 `mod` fromIntegral m1
        s11 = fromIntegral $ t2 `mod` fromIntegral m1
        s12 = fromIntegral $ t3 `mod` fromIntegral m1
        s20 = fromIntegral $ t4 `mod` fromIntegral m2
        s21 = fromIntegral $ t5 `mod` fromIntegral m2
        s22 = fromIntegral $ t6 `mod` fromIntegral m2

{-# INLINE restore #-}

jump :: Int -> MRGen -> MRGen
jump e g@(MRGen s10 s11 s12 s20 s21 s22)
  | e > 64    = error "Jump factor must be smaller than 64."
  | e == 0    = g
  | otherwise = MRGen t10 t11 t12 t20 t21 t22
  where m1' = fromIntegral m1 :: Word64
        m2' = fromIntegral m2 :: Word64
        v1 = floor <$> SV (s10, s11, s12)
        v2 = floor <$> SV (s20, s21, s22)
        (b1,b2) = jmtxs !! (e-1)
        w1 = vecTrOn m1' (fromIntegral <$> b1) v1
        w2 = vecTrOn m2' (fromIntegral <$> b2) v2
        SV (t10,t11,t12) = fromIntegral <$> w1
        SV (t20,t21,t22) = fromIntegral <$> w2

mtx1 :: JumpMatrix Word64
mtx1 = JM (0, 1, 0) (0, 0, 1) (m1 - a13n, a12, 0)

mtx2 :: JumpMatrix Word64
mtx2 = JM (0, 1, 0) (0, 0, 1) (m2 - a23n, 0, a21)

cntdn :: (a -> a) -> (a, Int) -> Maybe (a, (a, Int))
cntdn _ (_, 0) = Nothing
cntdn f (x, k) = Just (y, (y, k-1))
  where y = f x

jmtxs :: [(JumpMatrix Word32,JumpMatrix Word32)]
jmtxs = zip (map (fromIntegral <$>) xs) (map (fromIntegral <$>) ys)
  where n = 64
        mtx1' = fromIntegral <$> mtx1
        xs = unfoldr (cntdn (matSqrOn m1)) (mtx1',n)
        mtx2' = fromIntegral <$> mtx2
        ys = unfoldr (cntdn (matSqrOn m2)) (mtx2',n)

-- EOF


src/System/Random/MRG32k3a.hs
{- -*- mode: haskell; coding: utf-8-unix -*- -}
{-# LANGUAGE BangPatterns             #-}
{-# LANGUAGE DeriveDataTypeable       #-}
-- |
-- Module    : System.Random.MRG32k3a
-- Copyright : (c) 2020 Naoyuki MORITA
-- License   : BSD3
--
-- Maintainer  : naoyuki.morita@gmail.com
-- Stability   : experimental
-- Portability : portable
--
-- Pseudo-random number generation with MRG32k3a.
--
-- The generator type 'Gen' is an instance of 'RandomGen' type class, so
-- it can be used through 'RandomGen' intreface functions such like,
--
-- @
--   > let g = 'initialize' 12345
--   > let (x, g') = 'uniform' g :: (Word32, Gen) in x
--   3320887301
-- @
--
-- __Notice:__ MRG32k3a is originally designed to generate random numbers
-- following U(0,1). It DOES NOT generate exactly 32-bit information at a
-- time.
--
-- If you need U(0,1) random numbers, use 'uniform01' that generates a
-- random value efficiently by original MRG32k3a algorithm.
--
module System.Random.MRG32k3a
    (
    -- * Gen: Pseudo-Random Number Generators
      Gen
    , initialize

    -- * Unitility functions
    , uniform01

    -- * Seed: state management
    , Seed
    , fromSeed
    , save
    , restore

    -- * Stream jumping
    , jump
    ) where

import Data.Typeable (Typeable)
import Data.Word (Word32,Word64)
import Data.Bits ((.&.))

import System.Random

import System.Random.MRG.Internal
import System.Random.MRG32k3a.Internal

-- | The generator type.
newtype Gen = Gen (Double,Double,Double,Double,Double,Double)

instance RandomGen Gen where
  genWord32 = uniformW32
  split _ = error "Not yet implemented."

mrg32k3a_genRand :: Gen -> (Double,Gen)
mrg32k3a_genRand (Gen s@(_ ,s11,s12,_ ,s21,s22))
  = (v, Gen (s11,s12,t1,s21,s22,t2))
  where (t1,t2) = mrg32k3a_rec s
        v = if t1 <= t2 then t1 - t2 + m1f else t1 - t2
{-# INLINE mrg32k3a_genRand #-}

-- | Get a random value following U(0,1).
uniform01 :: Gen -> (Double,Gen)
uniform01 g = (w, g')
  where !w = v * norm
        (!v,!g') = mrg32k3a_genRand g
{-# INLINE uniform01 #-}

ub :: Word64
ub = v
  where !m1sq = m1 * m1
        !r = m1sq `mod` 4294967296
        !v = m1sq - r
{-# INLINE ub #-}

uniformW32 :: Gen -> (Word32,Gen)
uniformW32 g = if x >= ub
               then uniformW32 g1
               else (fromIntegral (x .&. 4294967295), g1)
  where (v0,g0) = mrg32k3a_genRand g
        (v1,g1) = mrg32k3a_genRand g0
        w0 = floor v0 :: Word64
        w1 = floor v1 :: Word64
        x = w0 * m1 + w1
{-# INLINE uniformW32 #-}

-- | Create a generator using given seed.
initialize :: Word32 -> Gen
initialize seed = Gen (s1,s1,s1,s2,s2,s2)
  where s' = fromIntegral seed
        s1 = fromIntegral $ s' `mod` m1
        s2 = fromIntegral $ s' `mod` m2
{-# INLINE initialize #-}

-- | An immutable snapshot of the state of a 'Gen'.
newtype Seed = Seed {
  -- | Convert seed into a 6-tuple of @Word32@.
  fromSeed :: (Word32,Word32,Word32,Word32,Word32,Word32)
  }
  deriving (Eq, Show, Typeable)

-- | Save the state of a 'Gen'. Saved state can be used by 'restore'.
save :: Gen -> Seed
save (Gen s) = Seed (fromT6 t)
  where t = floor <$> T6 s
{-# INLINE save #-}

-- | Create a new 'Gen' that mirrors the state of a saved 'Seed'.
restore :: Seed -> Gen
restore (Seed (t1,t2,t3,t4,t5,t6)) = Gen (fromT6 s)
  where m1' = fromIntegral m1
        m2' = fromIntegral m2
        t   = T6 (t1 `mod` m1', t2 `mod` m1', t3 `mod` m1',
                  t4 `mod` m2', t5 `mod` m2', t6 `mod` m2')
        s   = fromIntegral <$> t
{-# INLINE restore #-}

-- | Get a new generator jumps ahead by \(2^n\) steps from given generator.
--
-- @
--   > let g0 = 'initialize' 12345
--   > let g1 = 'jump' 20 g0
--   > let xs = unfoldr (Just . 'uniform01') g0
--   > let ys = unfoldr (Just . 'uniform01') g1
--   > take 10 $ drop 1048576 xs
--   [0.42963674510001276,0.10482156807623948,0.9889648413995019,0.785875227875553,0.9522150221887802,0.9792979233185687,0.8713777766671446,0.9231321178403405,0.13047652927672448,0.5395971153015737]
--   > take 10 $ ys
--   [0.42963674510001276,0.10482156807623948,0.9889648413995019,0.785875227875553,0.9522150221887802,0.9792979233185687,0.8713777766671446,0.9231321178403405,0.13047652927672448,0.5395971153015737]
-- @
jump :: Int -> Gen -> Gen
jump e (Gen (s10,s11,s12,s20,s21,s22))
  | e > 64 || e < 0 = error "Jump factor must be in the range of [0,64]."
  | otherwise       = Gen (t10,t11,t12,t20,t21,t22)
  where m1' = fromIntegral m1 :: Word64
        m2' = fromIntegral m2 :: Word64
        v1 = floor <$> SV (s10, s11, s12)
        v2 = floor <$> SV (s20, s21, s22)
        (b1,b2) = jmtxs !! e
        w1 = vecTrMod m1' (fromIntegral <$> b1) v1
        w2 = vecTrMod m2' (fromIntegral <$> b2) v2
        SV (!t10,!t11,!t12) = fromIntegral <$> w1
        SV (!t20,!t21,!t22) = fromIntegral <$> w2

-- EOF


src/System/Random/MRG63k3a.hs
{- -*- mode: haskell; coding: utf-8-unix -*- -}
{-# LANGUAGE BangPatterns             #-}
{-# LANGUAGE DeriveDataTypeable       #-}
-- |
-- Module    : System.Random.MRG63k3a
-- Copyright : (c) 2020 Naoyuki MORITA
-- License   : BSD3
--
-- Maintainer  : naoyuki.morita@gmail.com
-- Stability   : experimental
-- Portability : portable
--
-- Pseudo-random number generation with MRG63k3a.
--
-- The generator type 'Gen' is an instance of 'RandomGen' type class, so
-- it can be used through 'RandomGen' intreface functions such like,
--
-- @
--   > let g = 'initialize' 1234567
--   > let (x, g') = 'uniform' g :: (Word32, Gen) in x
--   2246106302
-- @
--
module System.Random.MRG63k3a
    (
    -- * Gen: Pseudo-Random Number Generators
      Gen
    , initialize

    -- * Unitility functions
    , uniform01

    -- * Seed: state management
    , Seed
    , fromSeed
    , save
    , restore

    -- * Stream jumping
    , jump
    ) where

import Data.Typeable (Typeable)
import Data.Int  (Int64)
import Data.Word (Word32,Word64)
import Data.Bits ((.&.))

import System.Random

import System.Random.MRG.Internal
import System.Random.MRG63k3a.Internal

-- | The generator type.
data Gen = Gen (Int64,Int64,Int64,Int64,Int64,Int64)

instance RandomGen Gen where
  genWord32 = uniformW32
  split _ = error "Not yet implemented."
  {-# INLINE genWord32 #-}

mrg63k3a_genRand :: Gen -> (Int64,Gen)
mrg63k3a_genRand (Gen s@(_ ,s11,s12,_ ,s21,s22))
  = (v, Gen (s11,s12,t1,s21,s22,t2))
  where (t1,t2) = mrg63k3a_rec s
        v = if t1 > t2 then t1 - t2 else t1 - t2 + m1
{-# INLINE mrg63k3a_genRand #-}

-- | Create a generator using given seed.
initialize :: Int64 -> Gen
initialize seed = Gen (s1,s1,s1,s2,s2,s2)
  where s1 = seed `mod` m1
        s2 = seed `mod` m2
{-# INLINE initialize #-}

-- | Get a random value following U(0,1).
uniform01 :: Gen -> (Double, Gen)
uniform01 gen = (w,gen')
  where (v,gen') = mrg63k3a_genRand gen
        !w = norm * fromIntegral v
{-# INLINE uniform01 #-}

ub :: Int64
ub = v
  where !r = m1 `mod` 4294967296
        !v = m1 - r
{-# INLINE ub #-}

uniformW32 :: Gen -> (Word32, Gen)
uniformW32 gen = go gen
  where go g = if x >= ub then go g' else (fromIntegral (x .&. 4294967295), g')
          where (x,g') = mrg63k3a_genRand g
{-# INLINE uniformW32 #-}

-- | An immutable snapshot of the state of a 'Gen'.
newtype Seed = Seed {
  -- | Convert seed into a 6-tuple of @Word64@.
  fromSeed :: (Word64,Word64,Word64,Word64,Word64,Word64)
  }
  deriving (Eq, Show, Typeable)

-- | Save the state of a 'Gen'. Saved state can be used by 'restore'.
save :: Gen -> Seed
save (Gen s) = Seed (fromT6 t)
  where t = fromIntegral <$> T6 s
{-# INLINE save #-}

-- | Create a new 'Gen' that mirrors the state of a saved 'Seed'.
restore :: Seed -> Gen
restore (Seed (t1,t2,t3,t4,t5,t6)) = Gen (fromT6 s)
  where m1' = fromIntegral m1
        m2' = fromIntegral m2
        t   = T6 (t1 `mod` m1', t2 `mod` m1', t3 `mod` m1',
                  t4 `mod` m2', t5 `mod` m2', t6 `mod` m2')
        s   = fromIntegral <$> t
{-# INLINE restore #-}

-- | Get a new generator jumps ahead by \(2^n\) steps from given generator.
--
-- @
--   >>> let g0 = initialize 1234567
--   >>> let g1 = jump 20 g0
--   >>> let xs = unfoldr (Just . uniform01) g0
--   >>> let ys = unfoldr (Just . uniform01) g1
--   >>> take 10 $ drop 1048576 xs
--   [0.8661862534637614,0.32299545841103733,0.7545203812431158,0.6829503373808978,0.7938803872652808,0.3453801891716969,0.199924353479426,0.7548724584148777,0.4533723232806143,0.855651888940174]
--   >>> take 10 ys
--   [0.8661862534637614,0.32299545841103733,0.7545203812431158,0.6829503373808978,0.7938803872652808,0.3453801891716969,0.199924353479426,0.7548724584148777,0.4533723232806143,0.855651888940174]
-- @
jump :: Int -> Gen -> Gen
jump e (Gen (s10,s11,s12,s20,s21,s22))
  | e > 64 || e < 0 = error "Jump factor must be in the range of [0,64]"
  | otherwise       = Gen (t10,t11,t12,t20,t21,t22)
  where m1' = fromIntegral m1
        m2' = fromIntegral m2
        v1  = fromIntegral <$> SV (s10, s11, s12)
        v2  = fromIntegral <$> SV (s20, s21, s22)
        (b1,b2) = jmtxs !! e
        w1  = vecTrModW64 m1' b1 v1
        w2  = vecTrModW64 m2' b2 v2
        SV (!t10,!t11,!t12) = fromIntegral <$> w1
        SV (!t20,!t21,!t22) = fromIntegral <$> w2

-- EOF


src/System/Random/MRG63k3a/Internal.hs
{- -*- mode: haskell; coding: utf-8-unix -*- -}
{-# LANGUAGE BangPatterns             #-}

module System.Random.MRG63k3a.Internal
    (
      norm
    , m1
    , m2

    , mrg63k3a_rec
    , jmtxs
    ) where

import System.Random.MRG.Internal

import Data.Int  (Int64)
import Data.Word (Word64)
import Data.List (unfoldr)

norm :: Double
norm = 1.0842021724855052e-19
{-# INLINE norm #-}

m1 :: Int64
m1 = 9223372036854769163
{-# INLINE m1 #-}

m2 :: Int64
m2 = 9223372036854754679
{-# INLINE m2 #-}

-- a12 * q12 + r12 = m1  ... (1)
a12 :: Int64
a12 = 1754669720
{-# INLINE a12 #-}

q12 :: Int64
q12 = 5256471877
{-# INLINE q12 #-}

r12 :: Int64
r12 = 251304723
{-# INLINE r12 #-}

-- a13n * q13 + r13 = m1 ... (2)
-- a13n * q13 = m1 - r13 ... (2)'
a13n :: Int64
a13n = 3182104042
{-# INLINE a13n #-}

q13 :: Int64
q13 = 2898513661
{-# INLINE q13 #-}

r13 :: Int64
r13 = 394451401
{-# INLINE r13 #-}

-- a21 * q21 + r21 = m2 ... (3)
a21 :: Int64
a21 = 31387477935
{-# INLINE a21 #-}

q21 :: Int64
q21 = 293855150
{-# INLINE q21 #-}

r21 :: Int64
r21 = 143639429
{-# INLINE r21 #-}

-- a23n * q23 + r23 = m2 ... (4)
a23n :: Int64
a23n = 6199136374
{-# INLINE a23n #-}

q23 :: Int64
q23 = 1487847900
{-# INLINE q23 #-}

r23 :: Int64
r23 = 985240079
{-# INLINE r23 #-}

mrg63k3a_rec :: (Int64,Int64,Int64,Int64,Int64,Int64) -> (Int64,Int64)
mrg63k3a_rec (s10,s11,_ ,s20,_ ,s22)
  = (t1,t2)
  where (h10,j10) = s10 `divMod` q13
        --  h10 * q13 + j10 = s10
        --  a13n * j10 - h10 * r13
        --    = a13n * (s10 - h10 * q13) - h10 * r13
        --    = a13n * s10 - h10 * (a13n * q13 + r13)
        --    { because of a13n * q13 + r13 = m1 (2), }
        --    = a13n * s10 - h10 * m1
        --
        p13  = a13n * j10 - h10 * r13
        -- When s10 = k * q13  (k = 1,2,..),
        --   a13n * s10 = a13n * k * q13
        --   { because of a13n * q13 = m1 - r13 (2)', }
        --      = k * (m1 - r13)
        --      = k * m1 - k * r13
        --      = (k-1) * m1 + (m1 - k * r13),
        -- while
        --   (h10, j10) = (k,0)  { because of s10 = k * q13 }
        --   a13n * j10 - h10 * r13 = - k * r13.
        -- Therefore,
        --   a13n * j10 - h10 * r13 + m1 = m1 - k * r13
        --      = a13n * s10 - (k-1) * m1
        --
        p13' = if p13 < 0 then p13 + m1 else p13
        (h11,j11) = s11 `divMod` q12
        --  h11 * q12 + j11 = s11
        --  a12 * j11 - h11 * r12
        --    = a12 * (s11 - h11 * q12) - h11 * r12
        --    = a12 * s11 - h11 * (a12 * q12 + r12)
        --    = a12 * s11 - h11 * m1
        --
        p12  = a12  * j11 - h11 * r12
        p12' = if p12  < 0 then p12  + m1 - p13' else p12 - p13'
        !t1  = if p12' < 0 then p12' + m1 else p12'
        (h20,j20) = s20 `divMod` q23
        --  h20 * q23 + j20 = s20
        --  a23n * j20 - h20 * r23
        --    = a23n * (s20 - h20 * q23) - h20 * r23
        --    = a23n * s20 - h20 * (a23n * q23 + r23)
        --    = a23n * s20 - h20 * m2
        --
        p23  = a23n * j20 - h20 * r23
        p23' = if p23 < 0 then p23 + m2 else p23
        (h22,j22) = s22 `divMod` q21
        --  h22 * q21 + j22 = s22
        --  a21 * j22 - h22 * r21
        --    = a21 * (s22 - h22 * q21) - h22 * r21
        --    = a21 * s22 - h22 * (a21 * q21 + r21)
        --    = a21 * s22 - h22 * m2
        --
        p21  = a21  * j22 - h22 * r21
        p21' = if p21  < 0 then p21  + m2 - p23' else p21 - p23'
        !t2  = if p21' < 0 then p21' + m2 else p21'
{-# INLINE mrg63k3a_rec #-}

mtx1 :: JumpMatrix Int64
mtx1 = JM (0, 1, 0) (0, 0, 1) (m1 - a13n, a12, 0)

mtx2 :: JumpMatrix Int64
mtx2 = JM (0, 1, 0) (0, 0, 1) (m2 - a23n, 0, a21)

cntdn :: (a -> a) -> (a, Int) -> Maybe (a, (a, Int))
cntdn _ (_, 0) = Nothing
cntdn f (x, k) = Just (y, (y, k-1))
  where y = f x

jmtxs :: [(JumpMatrix Word64, JumpMatrix Word64)]
jmtxs = zip xs ys
  where n = 64
        m1' = fromIntegral m1 :: Word64
        m2' = fromIntegral m2 :: Word64
        mtx1' = fromIntegral <$> mtx1
        xs = mtx1' : unfoldr (cntdn (matSqrModW64 m1')) (mtx1',n)
        mtx2' = fromIntegral <$> mtx2
        ys = mtx2' : unfoldr (cntdn (matSqrModW64 m2')) (mtx2',n)

-- EOF


src/System/Random/MRG63k3a/Stateful.hs
{- -*- mode: haskell; coding: utf-8-unix -*- -}
{-# LANGUAGE BangPatterns             #-}
{-# LANGUAGE DeriveDataTypeable       #-}
{-# LANGUAGE FlexibleContexts         #-}
{-# LANGUAGE FlexibleInstances        #-}
{-# LANGUAGE MultiParamTypeClasses    #-}
{-# LANGUAGE ScopedTypeVariables      #-}
{-# LANGUAGE TypeFamilies             #-}
-- |
-- Module    : System.Random.MRG63k3a.Stateful
-- Copyright : (c) 2020 Naoyuki MORITA
-- License   : BSD3
--
-- Maintainer  : naoyuki.morita@gmail.com
-- Stability   : experimental
-- Portability : portable
--
-- Pseudo-random number generation with MRG63k3a (monadic interface).
--
-- The generator type 'Gen' is an instance of 'StatefulGen' type class, so
-- it can be used through 'StatefulGen' intreface functions such like,
--
-- @
--   > gen \<- 'initialize' 1234567
--   > replicateM 10 ('uniformM' gen) :: IO [Word32]
--   [2246106302,1563963788,2439712072,3737154441,2667077669,767817191,747111673,2638409746,3331088863,4075662417]
--   > replicateM 10 ('uniformM' gen) :: IO [Word32]
--   [1456421684,2935764772,936846699,649810874,4215441082,311517124,1039486180,751453058,3053799799,1547236802]
-- @
--
module System.Random.MRG63k3a.Stateful
    (
    -- * Gen: Pseudo-Random Number Generators
      Gen
    , initialize

    -- ** Type helpers
    , GenIO
    , GenST

    -- * Unitility functions
    , uniform01M

    -- * Seed: state management
    , Seed
    , fromSeed
    ) where

import Control.Monad.Primitive (PrimMonad, PrimState, unsafeSTToPrim)
import Control.Monad.ST        (ST)
import qualified Data.Primitive.PrimArray as P
import Data.Typeable           (Typeable)
import Data.Int                (Int64)
import Data.Word
import Data.Bits               ((.&.),(.|.),shiftL)

import System.Random.MRG.Internal
import System.Random.MRG63k3a.Internal

import System.Random.Stateful

-- | State of the pseudo-random number generator. It uses mutable
-- state so same generator shouldn't be used from the different
-- threads simultaneously.
newtype Gen s = Gen (P.MutablePrimArray s Int64)

-- | A shorter name for PRNG state in the 'IO' monad.
type GenIO = Gen (PrimState IO)

-- | A shorter name for PRNG state in the 'ST' monad.
type GenST s = Gen (PrimState (ST s))

instance (s ~ PrimState m, PrimMonad m) => StatefulGen (Gen s) m where
  uniformWord32 = uniformW32
  uniformShortByteString n g = unsafeSTToPrim (genShortByteStringST n go)
    where go = do
            x0 <- uniformW32 g
            x1 <- uniformW32 g
            pure $ (fromIntegral x0 `shiftL` 32) .|. fromIntegral x1
  {-# INLINE uniformWord32 #-}
  {-# INLINE uniformShortByteString #-}

instance (PrimMonad m) => FrozenGen Seed m where
  type MutableGen Seed m = Gen (PrimState m)
  thawGen = restore
  freezeGen = save

readState :: (PrimMonad m) => P.MutablePrimArray (PrimState m) Int64 -> m (Int64,Int64,Int64,Int64,Int64,Int64)
readState ary = do
    s10 <- P.readPrimArray ary 0
    s11 <- P.readPrimArray ary 1
    s12 <- P.readPrimArray ary 2
    s20 <- P.readPrimArray ary 3
    s21 <- P.readPrimArray ary 4
    s22 <- P.readPrimArray ary 5
    return (s10,s11,s12,s20,s21,s22)
{-# INLINE readState #-}

writeState :: (PrimMonad m) => P.MutablePrimArray (PrimState m) Int64 -> (Int64,Int64,Int64,Int64,Int64,Int64) -> m ()
writeState ary (s10,s11,s12,s20,s21,s22) = do
    P.writePrimArray ary 0 s10
    P.writePrimArray ary 1 s11
    P.writePrimArray ary 2 s12
    P.writePrimArray ary 3 s20
    P.writePrimArray ary 4 s21
    P.writePrimArray ary 5 s22
{-# INLINE writeState #-}

mrg63k3a_genRand :: (PrimMonad m) => Gen (PrimState m) -> m Int64
mrg63k3a_genRand (Gen ary) = do
    s@(_,s11,s12,_,s21,s22) <- readState ary
    let (!t1,!t2) = mrg63k3a_rec s
        v = if t1 > t2 then t1 - t2 else t1 - t2 + m1
    writeState ary (s11,s12,t1,s21,s22,t2)
    return v
{-# INLINE mrg63k3a_genRand #-}

-- | Get a random value following U(0,1).
uniform01M :: (PrimMonad m) => Gen (PrimState m) -> m Double
uniform01M g = (* norm) <$> fromIntegral <$> mrg63k3a_genRand g
{-# INLINE uniform01M #-}

ub :: Int64
ub = m1 - r
  where !r = m1 `mod` 4294967296
{-# INLINE ub #-}

uniformW32 :: (PrimMonad m) => Gen (PrimState m) -> m Word32
uniformW32 gen = go
  where go = do
          v <- mrg63k3a_genRand gen
          if v >= ub then go else return (fromIntegral (v .&. 4294967295))
{-# INLINE uniformW32 #-}

-- | Create a generator using given seed.
initialize :: (PrimMonad m) => Int64 -> m (Gen (PrimState m))
initialize seed = do
    let s1 = seed `mod` m1
        s2 = seed `mod` m2
    ary <- P.newPrimArray 6
    writeState ary (s1,s1,s1,s2,s2,s2)
    return $ Gen ary
{-# INLINE initialize #-}

-- | An immutable snapshot of the state of a 'Gen'.
newtype Seed = Seed {
  -- | Convert seed into a 6-tuple of @Word64@.
  fromSeed :: (Word64,Word64,Word64,Word64,Word64,Word64)
  }
  deriving (Eq, Show, Typeable)

save :: (PrimMonad m) => Gen (PrimState m) -> m Seed
save (Gen ary) = do
    st <- readState ary
    return $ Seed $ fromT6 $ fromIntegral <$> (T6 st)
{-# INLINE save #-}

restore :: (PrimMonad m) => Seed -> m (Gen (PrimState m))
restore (Seed (t1,t2,t3,t4,t5,t6)) = do
    let m1' = fromIntegral m1
        m2' = fromIntegral m2
        st = T6 (t1 `mod` m1', t2 `mod` m1', t3 `mod` m1',
                 t4 `mod` m2', t5 `mod` m2', t6 `mod` m2')
    ary <- P.newPrimArray 6
    writeState ary (fromT6 (fromIntegral <$> st))
    return $ Gen ary
{-# INLINE restore #-}

-- EOF
